source/fe/mapping_q_generic.cc

```diff
@@ -3621,82 +3623,70 @@ namespace
         switch (structdim)
           {
           case 1:
           {
             const typename Triangulation<dim,spacedim>::line_iterator line = iter;
-            boundary->get_intermediate_points_on_line(line, points);
-            return;
+            boundary->get_intermediate_points_on_line(line, new_points);
+            break;
           }
           case 2:
           {
             const typename Triangulation<dim,spacedim>::quad_iterator quad = iter;
-            boundary->get_intermediate_points_on_quad(quad, points);
-            return;
+            boundary->get_intermediate_points_on_quad(quad, new_points);
+            break;
           }
           default:
             Assert(false, ExcInternalError());
-            return;
+            break;
           }
+        for (unsigned int i=0; i<new_points.size(); ++i)
+          points.push_back(new_points[i]);
       }
     else
       {
         std::vector<Point<spacedim> > sp(GeometryInfo<structdim>::vertices_per_cell);
         for (unsigned int i=0; i<sp.size(); ++i)
           sp[i] = iter->vertex(i);
-        get_intermediate_points(manifold, line_support_points, sp, points);
+        manifold.add_new_points(sp, weight_table, points);
       }
   }
 
 
-  /**
-   * Take a <tt>support_point_weights_on_hex(quad)</tt> and apply it to the vector
-   * @p a to compute the inner support points as a linear combination of the
-   * exterior points.
-   *
-   * The vector @p a initially contains the locations of the @p n_outer
-   * points, the @p n_inner computed inner points are appended.
-   *
-   * See equation (7) of the `mapping' report.
-   */
-  template <int spacedim>
-  void add_weighted_interior_points(const Table<2,double>   &lvs,
-                                    std::vector<Point<spacedim> > &a)
+
+  template <typename Iterator>
+  bool check_identical_manifolds_of_quads(const Iterator &)
   {
-    const unsigned int n_inner_apply=lvs.n_rows();
-    const unsigned int n_outer_apply=lvs.n_cols();
-    Assert(a.size()==n_outer_apply,
-           ExcDimensionMismatch(a.size(), n_outer_apply));
-
-    // compute each inner point as linear combination of the outer points. the
-    // weights are given by the lvs entries, the outer points are the first
-    // (existing) elements of a
-    for (unsigned int unit_point=0; unit_point<n_inner_apply; ++unit_point)
-      {
-        Assert(lvs.n_cols()==n_outer_apply, ExcInternalError());
-        Point<spacedim> p;
-        for (unsigned int k=0; k<n_outer_apply; ++k)
-          p+=lvs[unit_point][k]*a[k];
+    Assert(false, ExcNotImplemented());
+    return true;
+  }
 
-        a.push_back(p);
-      }
+  bool check_identical_manifolds_of_quads(const Triangulation<3,3>::cell_iterator &cell)
+  {
+    for (unsigned int f=0; f<GeometryInfo<3>::faces_per_cell; ++f)
+      if (&cell->face(f)->get_manifold() != &cell->get_manifold())
+        return false;
+    return true;
```

Comment:
I don't think you need to general template above if you write this function as
```
template <typename TriaIteratorType>
bool check_identical_manifolds_of_quads(const TriaIteratorType &cell)
￼{
  if (TriaIteratorType::dimension >= 2)
    for (unsigned int f=0; f<GeometryInfo<TriaIteratorType::dimension>::lines_per_cell; ++f)
      if (&cell->face(f)->get_manifold() != &cell->get_manifold())
        return false;
  return true;
}
```
That's because `GeometryInfo<TriaIteratorType::dimension>::lines_per_cell` should be defined for all dimensions.